### PR TITLE
No error logging on coroutine.resume (pull-request for issue #951)

### PR DESCRIPTION
**Pull request by bungle**
I hereby granted the copyright of the changes in this pull request to the authors of this lua-nginx-module project.

See more details on #951.

### CODE REVIEW

**Comment by agentzh** [comment]
@bungle Will you add some tests to the existing test suite to justify this change? Thank you!

**Comment by agentzh** [comment]
@bungle Also, please ensure the Travis CI checks are passing in this PR.

**Comment by bungle** [comment]
Yes, I will do that tomorrow.

**Comment by bungle** [comment]
@agentzh, it seems to pass the tests now. I don't think we need another tests for this as I did change the original tests to work with the patch, and they also test that now error is written to log on `coroutine.resume` where resumed `thread` errors.

File: t/091-coroutine.t [hunk 1]
```diff
@@ -483,8 +483,8 @@ done
 GET /lua
 --- response_body
 hello
---- error_log eval
-["stack traceback:", "coroutine 0:", "coroutine 1:", "coroutine 2:"]
+--- no_error_log
+[error]
```
Comment (agentzh) [inline t/091-coroutine.t]: Hmm, the indentation of this test case is to test the Lua land backtrace. After your change, we no longer have any tests designed for this purpose.

I'm not sure if silencing the error logging is a good idea. Maybe we can make it a lower log level like "warn" or even "info"?

File: t/091-coroutine.t [hunk 2]
```diff
@@ -4,7 +4,7 @@ use Test::Nginx::Socket::Lua;
 
 repeat_each(2);
 
-plan tests => repeat_each() * (blocks() * 3 + 5);
+plan tests => repeat_each() * blocks() * 3 + 4;
```
Comment (agentzh) [inline t/091-coroutine.t]: Better keep the parens otherwise we need to update this plan every time we change the repeat_each number. Also, the repeat each number is automatically increased on the "check leak" test mode of Test::Nginx::Socket, and your change will break the test plan in that test mode.

File: t/091-coroutine.t (hunk 1 above)
Comment (agentzh) [inline t/091-coroutine.t]: Also, better check the return values of `coroutine.resume` for better test coverage.

File: t/091-coroutine.t [hunk 3]
```diff
@@ -281,7 +281,7 @@ GET /lua
               end)
             end
 
-            N = 10 
+            N = 10
```
Comment (agentzh) [inline t/091-coroutine.t]: Better avoid unrelated changes to make your changeset atomic. It's better for version control.

File: t/091-coroutine.t [hunk 4]
```diff
@@ -1315,4 +1315,3 @@ co yield: 2
 
 --- no_error_log
 [error]
-
```
Comment (agentzh) [inline t/091-coroutine.t]: This is an unrelated change.

File: t/091-coroutine.t (hunk 3 above)
Comment (bungle) [inline t/091-coroutine.t]: I think my editor did it automatically.

File: t/091-coroutine.t (hunk 4 above)
Comment (bungle) [inline t/091-coroutine.t]: Yes, the editor did this automatically. I will add the new line back.

File: t/091-coroutine.t (hunk 1 above)
Comment (bungle) [inline t/091-coroutine.t]: `coroutine.resume` is a `pcall` and its purpose is to catch the errors aka silence the errors and let users handle them. I will make that test better so it checks the return value.

**Comment by bungle** [comment]
One thing though, is this code path only used with when user's code calls `coroutine.resume`, and not for say . Nginx managed light threads e.g. `ngx.thread` or for the main light thread.

File: src/ngx_http_lua_util.c [hunk 5]
```diff
@@ -1437,9 +1437,6 @@ ngx_http_lua_run_thread(lua_State *L, ngx_http_request_t *r,
 
             ctx->cur_co_ctx = next_coctx;
 
-            ngx_log_error(NGX_LOG_ERR, r->connection->log, 0,
-                          "lua coroutine: %s: %s\n%s", err, msg, trace);
```
Comment (agentzh) [inline src/ngx_http_lua_util.c]: As I mentioned earlier, should we just lower the log level here? Like `info`?

**Comment by agentzh** [comment]
@thibaultcha @doujiang24 What do you think about this?

File: src/ngx_http_lua_util.c (hunk 5 above)
Comment (bungle) [inline src/ngx_http_lua_util.c]: If we do so, this will be different than what Lua / LuaJIT does. They don't write anything. I mean, OpenResty doesn't log anything when wrapping code with `pcall`. If you want to log errors even when they are raised in a code wrapped with `pcall`, then I would consider having this logged as well, and of course we need to change `pcall` then as well. This would differ from what users' expect, though. My only concern is that I'm not sure if that removed log entry removes error logging also from some case where it shouldn't. That's why I asked do anyone of you know if it is only handling when users' code calls `coroutine.resume` (that's the only place where this logging should be silenced).

Comment (agentzh) [inline src/ngx_http_lua_util.c]: @bungle `pcall` is a more explicit `try/catch` in the Lua world while many Lua programmers may overlook the error handling in `coroutine.resume`.

Comment (bungle) [inline src/ngx_http_lua_util.c]: @agentzh, I agree, but that's how the Lua is. Those that use `coroutine.resume` should know that it is effectively a `pcall`. There is some history to this as `pcall` has not been fully compatible with coroutines (http://lua-users.org/wiki/PcallAndCoroutines).

**Comment by thibaultcha** [comment]
Sorry for the delay, I was on a trip.

I have spent some time testing this change as well as giving some thoughts about @bungle's concerns, and so far here is what I gathered:

- As we try to stay as close as possible as the standard Lua/LuaJIT's behavior, I believe this change to be aligned with this goal. We should not unexpectedly write to the error logs when Lua would not write to stderr itself.
- Ultimately, lowering the log level would be a compromise, but as with all compromises, it would leave both parties unsatisfied :)
- As for @bungle's concern in #951 about not providing a stacktrace to the `error()` call, I tried some cases on my side and it would seem that this behavior is particular to resty-cli, which executes all user code in a `pcall()`, as seen [here](https://github.com/openresty/resty-cli/blob/master/bin/resty#L353). I would be in favor of replacing this with an `xpcall` in order to provide a full stacktrace to the user and avoid confusions, but that is a different topic than this discussion.

So far with this patch, `error()` writes the appropriate errors and stacktraces from the various contexts I tried it in (`content`, `log`, `timer` and user threads) and indeed removes the logging in a coroutine. I feel like more testing would be necessary though.

**Comment by agentzh** [comment]
Okay, I'm fine with removing the logging in this case completely. We just need to ensure that a backtrace is still generated in other cases, i.e., outside a user `coroutine.resume()` call. In addition, we can avoid generating the backtrace in the `coroutine.resume()` case if we can since the backtrace is later discarded completely anyway and the generation is a little bit expensive.

**Comment by thibaultcha** [comment]
Done as part of #1239, closing this PR, thanks!
